extensions/pkg/controller/worker/genericactuator/actuator_reconcile.go

```diff
@@ -309,55 +324,115 @@ func (a *genericActuator) deployMachineDeployments(ctx context.Context, cluster
-		existingMachineDeployments := &machinev1alpha1.MachineDeploymentList{}
-		if err := a.client.List(ctx, existingMachineDeployments, client.InNamespace(worker.Namespace)); err != nil {
+	return retryutils.UntilTimeout(ctx, 5*time.Second, 3*time.Minute, func(ctx context.Context) (bool, error) {
+		var numHealthyDeployments, numUpdated, numAvailable, numUnavailable, numDesired, numberOfAwakeMachines int32
+
+		// Get the list of all machine deployments
+		machineDeployments := &machinev1alpha1.MachineDeploymentList{}
+		if err := a.client.List(ctx, machineDeployments, client.InNamespace(worker.Namespace)); err != nil {
+			return retryutils.SevereError(err)
+		}
+
+		// Get the list of all machine sets
+		machineSets := &machinev1alpha1.MachineSetList{}
+		if err := a.client.List(ctx, machineSets, client.InNamespace(worker.Namespace)); err != nil {
 			return retryutils.SevereError(err)
 		}
 
+		// map the owner reference to the machine sets
+		ownerReferenceToMachineSet := workerhelper.BuildOwnerToMachineSetMap(machineSets.Items)
+
 		// Collect the numbers of available and desired replicas.
-		for _, existingMachineDeployment := range existingMachineDeployments.Items {
+		for _, deployment := range machineDeployments.Items {
+			wantedDeployment := wantedMachineDeployments.GetDeploymentForName(deployment.Name)
+
 			// Filter out all machine deployments that are not desired (any more).
-			if !wantedMachineDeployments.HasDeployment(existingMachineDeployment.Name) {
+			if wantedDeployment == nil {
 				continue
 			}
 
+			machineSets := ownerReferenceToMachineSet[deployment.Name]
+
+			// use `wanted deployment` for these checks, as the existing deployments can be based on an outdated cache
+			alreadyExistingMachineDeployment := alreadyExistingMachineDeploymentNames.Has(wantedDeployment.Name)
+			newMachineClass := !alreadyExistingMachineClassNames.Has(wantedDeployment.ClassName)
+
+			if alreadyExistingMachineDeployment && newMachineClass {
+				rollingUpdate = true
+				a.logger.V(6).Info(fmt.Sprintf("Machine deployment %s is performing a rolling update", deployment.Name))
+				// Already existing machine deployments with a rolling update should have > 1 machine sets
+				if len(machineSets) <= 1 {
+					return retryutils.MinorError(errors.New(fmt.Sprintf("Waiting for the MachineControllerManager to create the machine sets for the machine deployment (%s/%s)...", deployment.Namespace, deployment.Name)))
+				}
+			}
+
+			// make sure that the machine set with the correct machine class for the machine deployment is already deployed
+			if machineSet := workerhelper.GetMachineSetWithMachineClass(wantedDeployment.Name, wantedDeployment.ClassName, ownerReferenceToMachineSet); machineSet == nil {
+				return retryutils.MinorError(errors.New(fmt.Sprintf("Waiting for the machine controller manager to create the updated machine set for the machine deployment (%s/%s)...", deployment.Namespace, deployment.Name)))
+			}
```

Comment:
same question here

Reply:
Another sanity check. Yes the NUMBER of machine sets might be two, but this does not garuantee that the required machine set with the correct class is present.
There could be two MachineSets, but those could be left-overs (e.g from the previous reconciliation & not garbage collected), or the  MCM is stuck.

If we do not detect this, this could lead to issues as we only check the status of the machine deployment (which could have a status indication the machine sets are fine, even though the correct machine set is not even created yet.)

Reply:
@prashanth26

Reply:
Thanks